# Reto Unidad 1
El primer paso es instalar en el entorno virtual ProyectoMasterIA creado en conda promp las librerias necesarias para la elaboración del ejercicio, que para este primer apartado son numpy para las operaciones matematicas y pandas para la limpieza de los datos

In [3]:
#Se importan las lbrearias a utilizar
import numpy as np
import pandas as pd
print("ok")

ok


Posteriormente se pasa el archivo csv a un dataframe con el objetivo de analizar la estructura de la información del archivo que esta separado por ,

In [4]:

df=pd.read_csv('ventas_techventas.csv',
            sep=",", 
            encoding="latin-1")
df.head()

,order_id,fecha,cliente_id,cliente_nombre,region,producto,categoria,cantidad,precio_unitario,descuento,vendedor
0,1001,2024-01-05,C001,TechCorp SA,Norte,Laptop Pro 15,Electronica,2.0,1200.0,0.10,Ana L�ï¿
1,1002,2024-01-07,C002,Innovatek,Sur,Mouse Inalambrico,Perifericos,15.0,25.0,0.00,Carlos Ruiz
2,"1003,2024-01-10,C003,DataSolutions,Centro,Moni...",NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,1004,2024-01-12,C001,TechCorp SA,Norte,Teclado Mecanico,Perifericos,10.0,85.0,0.10,Luis Mora
4,1005,2024-01-15,C004,CloudNet,Este,Laptop Pro 15,Electronica,3.0,1200.0,0.15,Carlos Ruiz


In [5]:
#Aqui se toman 2 columnas del dataframe y los convertimos en 2 arreglos de numpy: cantidad y precio y luego se imprimen por pantalla para ver como se guardaron

cantidad=df['cantidad'].values
precio=df['precio_unitario'].values

print(cantidad)

print("precios-------------")
print(precio)

[ 2. 15. nan 10.  3. 20. nan  8. 12.  1. nan 30.  5.  3. 25.  2. nan 18.
 12.  2.  4. 30.  8. nan  4. 20. 15.  6.  1. 10. nan 22.  5. 25.  2. nan
 12. 18. nan  3. 15.  5.  2.  3. 20.  6. 18. -1.  4. 25. nan  2. 20. 35.
 nan  2.  7.  9. 20. nan]
precios-------------
[1200.   25.   nan   85. 1200.   95.   nan  120.   95. 1200.   nan   25.
   85.  120.   95. 1200.   nan   95.   25. 1200.  120.   95.   85.   nan
 1200.   25.   95.  120. 1200.   85.   nan   95. 1200.   25.  120.   nan
   85.   95.   nan 1200.   25.  120.  350. 1200.   95.   85.   25. 1200.
  120.   95.   nan 1200.   25.   95.   nan 1200.  120.   85.   95.   nan]


In [6]:
#Aqui se detectan cuántos valores faltantes (NaN) hay en las columnas cantidad y precio_unitario.
print(df[['cantidad', 'precio_unitario']].isna().sum())

cantidad           11
precio_unitario    11
dtype: int64


In [ ]:
#Aqui se calcula el producto punto entre los dos arreglos cantidad y precio, pero su resultado es nan debido a que los arreglos todavía tienen valores nulos

ingresos_brutos = np.dot(cantidad,precio)

print(ingresos_brutos)

nan


Antes de realizar cualquier imputación o cambio en el archivo csv, se crea una copia independiente para poder modificar el archivo sin afectar el DataFrame original

In [ ]:
df_ventas=df.copy() #se crea la copia

#Se reemplazan los valores faltantes (NaN) por 0 en las columnas cantidad y precio_unitario porque el 0 
#  significa ausencia de cantidad o de precio.

df_ventas['cantidad']=df_ventas['cantidad'].fillna(0)
df_ventas['precio_unitario']=df_ventas['precio_unitario'].fillna(0)

In [ ]:
#Una vez imputados los valores nulos, se vuelve a verificar si la base de datos sigue teniendo valores nulos
print(df_ventas[['cantidad', 'precio_unitario']].isna().sum())

cantidad           0
precio_unitario    0
dtype: int64


In [ ]:
#Se obtienen las columnas cantidad y precio_unitario del DataFrame  y se convierten en arreglos de numpy para poder realizar operaciones matemáticas

cantidades=df_ventas['cantidad'].values
precios=df_ventas['precio_unitario'].values

#Después se realiza una multiplicación elemento a elemento donde cantidad se multiplica por su precio correspondiente y el resultado es un arreglo donde cada posición representa el ingreso generado por un producto.

ingresos_de_elemento=(cantidades*precios)
print(ingresos_de_elemento)

#Se calcula el producto punto que multiplica los elementos correspondientes y luego suma todos los resultados para obtener el ingreso bruto total.
ingresos_brutos_totales = np.dot(cantidades,precios)

print()
print(f" los ingresos totales son de ${ingresos_brutos_totales}")


#Se suman todos los ingresos calculados mediante la multiplicación elemento a elemento y se compara con el resultado de producto punto, donde si son iguales, arroja true
print("\nSuma de la multiplicación elemento a elemento:")
print(ingresos_de_elemento.sum())

print("\n¿Son iguales?")
print(ingresos_de_elemento.sum() == ingresos_brutos_totales)

[ 2400.   375.     0.   850.  3600.  1900.     0.   960.  1140.  1200.
     0.   750.   425.   360.  2375.  2400.     0.  1710.   300.  2400.
   480.  2850.   680.     0.  4800.   500.  1425.   720.  1200.   850.
     0.  2090.  6000.   625.   240.     0.  1020.  1710.     0.  3600.
   375.   600.   700.  3600.  1900.   510.   450. -1200.   480.  2375.
     0.  2400.   500.  3325.     0.  2400.   840.   765.  1900.     0.]

 los ingresos totales son de $73855.0

Suma de la multiplicación elemento a elemento:
73855.0

¿Son iguales?
True


In [ ]:
#Aqui se calculan estadísticas descriptivas de la columna precios para entender cómo se distribuyen los datos.
print(np.mean(precios))
print(np.median(precios))
print(np.std(precios))
print(np.percentile(precios,[25,75]))

314.0833333333333
95.0
469.08909216575154
[ 25. 120.]


# Explicación estadistica
Los resultados de las estadísticas descriptivas del bloque anterior muestran cómo se distribuyen los precios del catálogo donde la media (314.08) es bastante alta en comparación con la mayoría de los valores, lo que indica que está influenciada por productos más costosos, sin embargo, la mediana (95.0) representa mejor el precio típico, ya que divide el catálogo en dos partes iguales y refleja que la mayoría de los productos se encuentran alrededor de ese valor. La desviación estándar (469.09) es alta, lo que indica que existe una gran variabilidad entre los precios, es decir, no todos los productos tienen valores similares y hay diferencias importantes dentro del catálogo y finalmente, los percentiles [25, 120] muestran que el 25% de los productos tienen precios iguales o menores a 25, mientras que el 75% están por debajo o igual a 120, lo que confirma que la mayoría de los productos son relativamente económicos y que solo unos pocos tienen precios más altos que elevan el promedio.

# Reflexión CRISP-DM
Mapeo del reto a CRISP-DM

1. Comprensión del negocio:
Se entiende el objetivo del reto: analizar ventas para obtener ingresos, precios y comportamiento del catálogo donde se define qué se quiere responder (por ejemplo, ingresos totales y distribución de precios).

2. Comprensión de los datos:
Se explora el dataset (cantidad, precio_unitario, fechas, productos), se revisan valores faltantes con isna() y se analizan estadísticas como media, mediana y percentiles.

3. Preparación de los datos:
Se limpian los datos reemplazando valores nulos con fillna(0), se seleccionan columnas relevantes y se convierten a arrays con .values para trabajar con numpy.

4. Modelado:
Se realizan operaciones matemáticas con numpy como la multiplicación elemento a elemento y el producto punto (np.dot) para calcular ingresos por producto y totales.

5. Evaluación:
Se comparan resultados como np.dot(cantidades, precios) vs (cantidades * precios).sum() para verificar consistencia y validar que los cálculos son correctos.

6. Despliegue:
En este contexto educativo no hay despliegue real, pero los resultados podrían usarse en reportes o dashboards para toma de decisiones.

Reflexión:

La fase más importante en este caso es la comprensión y preparación de los datos, porque si los datos están incompletos o mal tratados (por ejemplo, valores nulos o tipos incorrectos), todos los cálculos posteriores pueden ser incorrectos ya que por ejemplo el modelado con numpy es sencillo, pero su resultado depende completamente de la calidad de los datos, por lo que sin una buena limpieza y entendimiento previo, las métricas como ingresos o estadísticas del catálogo pueden llevar a conclusiones equivocadas. 